src/nomad_simulations/physical_property.py

```diff
@@ -0,0 +1,231 @@
+# limitations under the License.
+#
+
+from typing import Any, Optional
+
+from nomad import utils
+from nomad.datamodel.data import ArchiveSection
+from nomad.metainfo import (
+    Quantity,
+    SubSection,
+    SectionProxy,
+    Reference,
+    Section,
+    Context,
+    MEnum,
+)
+from nomad.metainfo.metainfo import DirectQuantity, Dimension, _placeholder_quantity
+from nomad.datamodel.metainfo.basesections import Entity
+
+from .variables import Variables
+from .numerical_settings import SelfConsistency
+
+
+# We add `logger` for the `PhysicalProperty.variables_shape` method
+logger = utils.get_logger(__name__)
+
+
+class PhysicalProperty(ArchiveSection):
+    """
+    A base section used to define the physical properties obtained in a simulation, experiment, or in a post-processing
+    analysis. Besides its `value`, it also contains:
+        - Quantities that are string identifiers: `source`, `type`, `label`
+        - Quantities for calculating the shape of the physical property: `shape`, `variables`
+        - Quantities for referencing and establishing the character of a physical property: `outputs_ref`, `entity_ref`, `self_consistent_ref`,
+        `is_derived`, `is_scf_converged`
+    """
+
+    name = Quantity(
+        type=str,
+        description="""
+        Name of the physical property. Example: `'ElectronicBandGap'`.
+        """,
+    )
+
+    source = Quantity(
+        type=MEnum('simulation', 'measurement', 'analysis'),
+        default='simulation',
+        description="""
+        Source of the physical property. Example: an `ElectronicBandGap` can be obtained from a `'simulation'`
+        or in an `'measurement'`.
+        """,
+    )
+
+    type = Quantity(
+        type=str,
+        description="""
+        Type categorization of the physical property. Example: an `ElectronicBandGap` can be `'direct'`
+        or `'indirect'`.
+        """,
+        # ! add more examples in the description
+    )
+
+    label = Quantity(
+        type=str,
+        description="""
+        Label for additional classification of the physical property. Example: an `ElectronicBandGap`
+        can be labeled as `'DFT'` or `'GW'` depending on the methodology used to calculate it.
+        """,
+        # ! add more examples in the description
+    )
+
+    rank = DirectQuantity(
+        type=Dimension,
+        shape=['0..*'],
+        default=[],
+        name='rank',
+        description="""
+        Rank of the tensor describing the physical property. This quantity is stored as a Dimension:
+            - scalars (tensor rank 0) have `rank=[]` (`len(rank) = 0`),
+            - vectors (tensor rank 1) have `rank=[a]` (`len(rank) = 1`),
+            - matrices (tensor rank 2), have `rank=[a, b]` (`len(rank) = 2`),
+            - etc.
+        """,
+    )
+
+    variables = SubSection(sub_section=Variables.m_def, repeats=True)
+
+    value: Quantity = _placeholder_quantity
```

Comment:
was just wondering how this works for the custom property? I guess it doesn't matter if value can be any datatype. I was thinking at first that it was always a float, but now I'm not sure :thinking:

Reply:
Yeah, I think in any case `value` has to be set up in the custom property you want to define.

Reply:
Shame that you can't add a description here.
Though, `property()` upon which `_placeholder_quantity` is based does allow for `doc`

Reply:
After our meeting with the interoperability TF, I will leave this to talk with Lauri, Markus and Theodore next week.

My problem here was that, when instantiating `value` as a Quantity with type, description, etc, then I cannot assign shape in a dynamic way (and similarly, unit will not be possible to be done like this), and rather it gets assign when initiating for the first time (even if it is a None, I cannot assign it a posteriori). I think this could be fixed if these attrs could be assigned `Any` for any potential value, but I will hear from Area D first about this.

So if you all agree, I keep this as it is now, and next week we will take care of it.

Reply:
Sure, I indeed think it's advisable to consult Area D here.

Whether this should already be merged depends on any deadlines. From my side, at least, don't feel pressured.
This approach is foundational enough, that it warrants a proper implementation: we don't want a plugin built on "good-enough-for-now" code.